[#64132] Allow `TreeView` nodes to be anchors or buttons

camertron opened the pull request:
### What are you trying to accomplish?
<!-- Provide a description of the changes. -->

This PR allows `TreeView` nodes to be anchors or buttons to enable using the `TreeView` component for navigation.

### Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

### Integration
<!-- Does this change require any updates to code in production? -->

No changes necessary in production.

#### List the issues that this change affects.
<!--Every code change must address _at least 1_ issue. Fixes a bug, completes a task, every change
      should have a corresponding issue listed here. If one does not already exist, create one. -->

https://community.openproject.org/projects/design-system/work_packages/64132/activity

#### Risk Assessment
  <!-- Please select from one of the following and detail why this level was chosen -->

- [x] **Low risk** the change is small, highly observable, and easily rolled back.
- [ ] **Medium risk** changes that are isolated, reduced in scope or could impact few users. The change will not impact library availability.
- [ ] **High risk** changes are those that could impact customers and SLOs, low or no test coverage, low observability, or slow to rollback.

### Accessibility
<!--
  You may remove this section and the "Accessibility" heading above _only_ if the changes in this pull request do not impact UI. Delete all those that don't apply.
  If there are any accessibility-related updates, please describe them here.
-->
- **No new axe scan violation** - This change does not introduce any new [axe scan](https://thehub.github.com/epd/engineering/dev-practicals/frontend/accessibility/readiness-routine/development/#axe-scans) violations.

### Merge checklist

- [x] Added/updated tests
- [x] Added/updated documentation
- [x] Added/updated previews (Lookbook)
- [x] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

HDinger reviewed (changes requested):
Hi @camertron 

looks already pretty solid 👍  I found one small issue: The area in which the cursor changes is on the left currently larger than the actual link. This creates the impression that I can click although nothing happens when I do:

<img width="461" alt="Bildschirmfoto 2025-05-21 um 13 16 13" src="https://github.com/user-attachments/assets/056ee170-99d9-49f9-8343-0f3069289e15" />

Further, we [discussed](https://community.openproject.org/meetings/5193#item-14014) the open questions from the ticket today. Parimal will write a detailed answer in the ticket, but these are basically the two points which require changes in your PR:

* It would be cool to have an underline effect for anchors to have at least some visual clue that this is an interactive element
* We should not mix anchors, buttons and divs for nodes/leaves in the same implementation. It should be the same tag for all leaves/nodes. Parimal would like that to be forbidden on a component level.

camertron commented:
Hey @HDinger, thanks for your feedback!

> It would be cool to have an underline effect for anchors to have at least some visual clue that this is an interactive element.

Done! I added an underline specifically to the hover state, since adding it to the normal state didn't look quite right to me. I think that's in-line with the rest of Primer, but please let me know if you disagree.

> We should not mix anchors, buttons and divs for nodes/leaves in the same implementation. It should be the same tag for all leaves/nodes. Parimal would like that to be forbidden on a component level.

That makes sense. To implement this, I removed the `tag:` option from `Node` and added a `node_variant:` option to the top-level `TreeView` constructor. In other words, to make every node an `<a>` element, you write:

```ruby
TreeView.new(node_variant: :anchor)
```

Let me know what you think.

> The area in which the cursor changes is on the left currently larger than the actual link. This creates the impression that I can click although nothing happens when I do

I tried looking into this today, and I have a few thoughts:

1. For anchor nodes, I wonder if the newly added underline is enough of a visual indication of where you can click that we don't need to make the empty space you indicated clickable.
2. For button nodes, things are more complicated. The space to the left of the button is usually where the expand/collapse arrow would go, but for leaf nodes it's an empty grid space. I tried putting a `<div>` in there and hooked up an event listener on it, but JavaScript doesn't really provide a way of forwarding those events to the `<button>`. We could make an exception for the `click` event and simply call `buttonElement.click()`, but not only would some of the original event information be lost, it doesn't account for all the other events that buttons emit. One potential solution might be to add an underline to the button text on hover like we're doing for anchors, but then the button looks like a link. Maybe that's ok tho.

HDinger commented:
Hi @camertron 

Parimal and I finally discussed this.

>    1. For anchor nodes, I wonder if the newly added underline is enough of a visual indication of where you can click that we don't need to make the empty space you indicated clickable.

>    The space to the left of the button is usually where the expand/collapse arrow would go, but for leaf nodes it's an empty grid space. I tried putting a <div> in there and hooked up an event listener on it, but JavaScript doesn't really provide a way of forwarding those events to the <button>. We could make an exception for the click event and simply call buttonElement.click(), but not only would some of the original event information be lost, it doesn't account for all the other events that buttons can emit. 

Actually, what I had in mind was to change the `cursor` behaviour in CSS only for both buttons and anchors. Currently, it is set to `pointer` for the whole `TreeViewItemContainer`. We could change that to set the cursor for `TreeViewItemToggle` and `TreeViewItemContent` only. I did not check that in detail, but at least for the link preview 

<img width="1196" alt="Bildschirmfoto 2025-05-22 um 14 06 59" src="https://github.com/user-attachments/assets/3e516a03-c9fd-47d5-bdf5-921bf781544e" />

>   One potential solution might be to add an underline to the button text on hover like we're doing for anchors, but then the button looks like a link. Maybe that's ok tho.

With regards to the Button highlighting, we decided to go with the approach that you suggested and also show an underline effect on hover for buttons similar to Links. The reason for this is that users can't visually tell the difference between an anchor and a button. They just need to know that there is an action behind.

​

camertron commented:
> Actually, what I had in mind was to change the cursor behaviour in CSS only for both buttons and anchors. Currently, it is set to pointer for the whole TreeViewItemContainer. We could change that to set the cursor for TreeViewItemToggle and TreeViewItemContent only.

Ah ok that makes sense, I'll make the change 😎 

> With regards to the Button highlighting, we decided to go with the approach that you suggested and also show an underline effect on hover for buttons similar to Links. The reason for this is that users can't visually tell the difference between an anchor and a button. They just need to know that there is an action behind.

Ok perfect, I had the same thought. I'll make that change as well 👍

camertron commented:
Ok @HDinger, I think all the behavior we have discussed so far has been implemented 👍

HDinger reviewed (approved):
Thanks a lot @camertron 🙇 Lgtm 🌟
